### PR TITLE
[upstream-10.0][SZGH3129][SZPR3496][ADD] improve code add 2 function to be override

### PR DESCRIPTION
**Pull request by Reinhard-sheng**
@Elico-Corp/odoo-reviewers 
szmig-odoo issue 3129

### CODE REVIEW

File: base_report_docx/report/report_docx.py
Comment (elicoidal) [inline base_report_docx/report/report_docx.py]: why not directly add here the function. it could be useful for other cases, no?

File: base_report_docx/report/report_docx.py
Comment (elicoidal) [inline base_report_docx/report/report_docx.py]: not sure we should do it like that: too specific.
This function should be added as a super to the render function I think (in specific module)
cc @victormartinelicocorp

**Comment by leo8tony** [comment]
@Elico-Corp/odoo-reviewers Please review ASAP, IT impact repacking label printing

File: base_report_docx/report/report_docx.py
Comment (JoJoJoJoJoJoJo) [inline base_report_docx/report/report_docx.py:80]:
```suggestion
        raise NotImplementedError
```
The error should not be muted.

**Review by JoJoJoJoJoJoJo — CHANGES_REQUESTED** [review]
Small details

File: base_report_docx/report/report_docx.py
Comment (JoJoJoJoJoJoJo) [inline base_report_docx/report/report_docx.py:250]:
```suggestion
        raise NotImplementedError
```

File: base_report_docx/report/report_docx.py
Comment (Reinhard-sheng) [reply base_report_docx/report/report_docx.py]: No, other report not have the same logic. can't put it here.

File: base_report_docx/report/report_docx.py
Comment (Reinhard-sheng) [reply base_report_docx/report/report_docx.py]: Agree, but it will be a big change for the base module and also the module inherited.

File: base_report_docx/report/report_docx.py
Comment (martintorresvm) [reply base_report_docx/report/report_docx.py]: @elicoidal I've talked to @Reinhard-sheng and explain that would need about 8h to refactor this and the others functions. So feels like he can't be on time.

@Reinhard-sheng I think you can go ahead and next time just give us feedback if you have any concern.

cc @leo8tony

File: base_report_docx/report/report_docx.py
Comment (martintorresvm) [reply base_report_docx/report/report_docx.py]: please do it. Otherwise we wouldn't ask for it

**Comment by elicoidal** [comment]
@JoJoJoJoJoJoJo @victormartinelicocorp I let you drive this change. Remember this module is public and needs to stay generic enough.

**Comment by mengzeyang** [comment]
@Reinhard-sheng Please fix

**Comment by Reinhard-sheng** [comment]
Sorry, I tried but confused.
Most of this code is not able to inherit. So I add function to override.
For example:
```
    def create_source_docx(self, cr, uid, ids, context=None):
        data = self.generate_docx_data(cr, uid, ids, context)
        original_folder_name = '/tmp/docx_to_pdf/'
        tmp_folder_name = original_folder_name + \
            str(int(time.time())) + \
            str(int(1000 + random.random() * 1000)) + '/'

        output_type = self._get_output_type(cr, uid, context)
        output_report = {
            'pdf': 'report.pdf',
            'docx': 'report.docx'
        }
        self._delete_temp_folder(tmp_folder_name)
        self._create_temp_folder(tmp_folder_name)

        self._generate_reports(
            cr, uid, context, tmp_folder_name, data,
            output_type, output_report)

        report = self._get_convert_file(
            tmp_folder_name, output_report[output_type])
        self.save_pdf_to_share(cr, uid, ids, report, tmp_folder_name, context)
        self._delete_temp_folder(tmp_folder_name)

        return (report, output_type)
```
the params in function `save_pdf_to_share` like `tmp_folder_name` and `report` only exists in there.
The `tmp_folder_name` is generated by timestamp and random, and for the report it will return to odoo.
And for the function add_barcode_for_doc, the render function is by lib `from docxtpl import DocxTemplate`, if we want to inherit, we have to inherit the DocxTemplate? Is that we want?

**Comment by Reinhard-sheng** [comment]
This module in V7 is not have save pdf and add barcode in the base module. Function for SCIP to printed by printer is special.
Could you give some suggestions?
There are a lot of code about IO and sys operation, like cp file and delete file. 
For example, we have function like this:
```
def A():
    io_file = create_file()
    additional_operation_for_file()
    delete_file()
    return io_file
```
we have to put the function `additional_operation_for_file` here, right?

**Comment by Reinhard-sheng** [comment]
@Elico-Corp/odoo-reviewers 
need help

**Comment by elicoidal** [comment]
@JoJoJoJoJoJoJo @victormartinelicocorp @seb-elico @joycehuan please help in this

File: base_report_docx/report/report_docx.py
Comment (seb-elico) [inline base_report_docx/report/report_docx.py]: change to `on_delete_output_file(file)`

where `file = tmp_folder_name + output_report[output_type]`

and change the signature of: `_get_convert_file(file)`

File: base_report_docx/report/report_docx.py
Comment (seb-elico) [reply base_report_docx/report/report_docx.py]: Rename to `on_doc_render(doc, context, data)`

And add a comment to say that this is a hook to perform additional operations on the DOC file before it's rendered.

**Review by seb-elico — CHANGES_REQUESTED** [review]

File: base_report_docx/report/report_docx.py
Comment (seb-elico) [reply base_report_docx/report/report_docx.py]: Also, adapt the comment to explain that this is a hook that can be used to copy the file to the HD (definitely do not mention shared folder, the shared folder is a deployment topic)

**Comment by seb-elico** [comment]
@Reinhard-sheng Be careful with naming of PR. [i3496] means ERP issue with ID 3496 (and for people not following the rule, it could mean GH issue ID 3496), but definitely it's not a PR ID!

**Comment by Reinhard-sheng** [comment]
@Elico-Corp/odoo-reviewers 
code updated, rename the function, add comments

**Review by JoJoJoJoJoJoJo — APPROVED** [review]
LGTM

**Comment by mengzeyang** [comment]
@seb-elico Please review.

**Comment by seb-elico** [comment]
> @seb-elico Please review.

@mengzeyang As a reminder, when the comments of a reviewer have been addressed, any other reviewer is allowed to approve the PR, as soon as there are 2 consecutive approval without any commit in between. It is not necessary to wait for the other reviewers to review it again.

CC @victormartinelicocorp @JoJoJoJoJoJoJo @joycehuan @leo8tony @elicoidal

File: base_report_docx/report/report_docx.py
Comment (seb-elico) [inline base_report_docx/report/report_docx.py]: Any operation can be performed in this hook. This comment is way too specific. Replace with:
"Override this method to perform operations on the output file before it gets deleted. For instance, you can copy it to another location in the file system."

File: base_report_docx/report/report_docx.py
Comment (seb-elico) [inline base_report_docx/report/report_docx.py]: Replace with the same comment as the one above:
"Override this function to perform additional operations on the DOC file before it's rendered."

**Review by seb-elico — CHANGES_REQUESTED** [review]

**Comment by Reinhard-sheng** [comment]
@seb-elico 
Updated comments.

**Review by joycehuan — APPROVED** [review]

**Comment by seb-elico** [comment]
> @seb-elico
Updated comments.

@Reinhard-sheng Please check my message above. When you fix the comments, please @Elico-Corp/odoo-reviewers , not only the person who asked for the changes. Any member of @Elico-Corp/odoo-reviewers can approve the modifications.

**Pull request merged**
